Core/GDCore/Project/ObjectFolderOrObject.cpp

```diff
@@ -0,0 +1,236 @@
+      });
+  if (it == children.end()) return;
+
+  children.erase(it);
+}
+
+void ObjectFolderOrObject::MoveObjectFolderOrObjectToAnotherFolder(
+    gd::ObjectFolderOrObject& objectFolderOrObject,
+    gd::ObjectFolderOrObject& newParentFolder,
+    std::size_t newPosition) {
+  if (!newParentFolder.IsFolder()) return;
+  if (newParentFolder.IsADescendantOf(objectFolderOrObject)) return;
+
+  std::vector<std::unique_ptr<gd::ObjectFolderOrObject>>::iterator it =
+      find_if(children.begin(),
+              children.end(),
+              [&objectFolderOrObject](std::unique_ptr<gd::ObjectFolderOrObject>&
+                                          childObjectFolderOrObject) {
+                return &(*childObjectFolderOrObject) == &objectFolderOrObject;
+              });
+  if (it == children.end()) return;
+
+  std::unique_ptr<gd::ObjectFolderOrObject> objectFolderOrObjectPtr =
+      std::move(*it);
+  children.erase(it);
+
+  objectFolderOrObjectPtr->parent = &newParentFolder;
+  newParentFolder.children.insert(
+      newPosition < newParentFolder.children.size()
+          ? newParentFolder.children.begin() + newPosition
+          : newParentFolder.children.end(),
+      std::move(objectFolderOrObjectPtr));
+}
+
+void ObjectFolderOrObject::SerializeTo(SerializerElement& element) const {
+  if (IsFolder()) {
+    element.SetAttribute("folderName", GetFolderName());
+    if (children.size() > 0) {
+      SerializerElement& childrenElement = element.AddChild("children");
+      childrenElement.ConsiderAsArrayOf("objectFolderOrObject");
+      for (std::size_t j = 0; j < children.size(); j++) {
+        children[j]->SerializeTo(
+            childrenElement.AddChild("objectFolderOrObject"));
+      }
+    }
+  } else {
+    element.SetAttribute("objectName", GetObject().GetName());
+  }
+}
+
+void ObjectFolderOrObject::UnserializeFrom(
+    gd::Project& project,
+    const SerializerElement& element,
+    gd::ObjectsContainer& objectsContainer) {
+  children.clear();
+  gd::String potentialFolderName = element.GetStringAttribute("folderName", "");
+
+  if (!potentialFolderName.empty()) {
+    object = nullptr;
+    folderName = potentialFolderName;
+
+    if (element.HasChild("children")) {
+      const SerializerElement& childrenElements =
+          element.GetChild("children", 0);
+      childrenElements.ConsiderAsArrayOf("objectFolderOrObject");
+      for (std::size_t i = 0; i < childrenElements.GetChildrenCount(); ++i) {
+        std::unique_ptr<ObjectFolderOrObject> childObjectFolderOrObject =
+            make_unique<ObjectFolderOrObject>();
+        childObjectFolderOrObject->UnserializeFrom(
+            project, childrenElements.GetChild(i), objectsContainer);
+        childObjectFolderOrObject->parent = this;
+        children.push_back(std::move(childObjectFolderOrObject));
+      }
+    }
+  } else {
+    folderName = "";
+    object =
+        &objectsContainer.GetObject(element.GetStringAttribute("objectName"));
```

Comment:
What happens if the object does not exist? I would protect against it by showing a `gd::LogWarning` or `gd::LogError`

Reply:
and keep a nullptr to show that "there is an issue here, hope the rest of the code will handle this properly".

Reply:
Good idea!